compiler/hash-ir/src/constant.rs

```diff
@@ -0,0 +1,537 @@
+            return Err(());
+        };
+
+        match char::from_u32(val.try_into().unwrap()) {
+            Some(c) => Ok(c),
+            None => Err(()),
+        }
+    }
+}
+
+impl From<IntConstant> for Scalar {
+    fn from(value: IntConstant) -> Self {
+        let size = value.size();
+
+        match value.value {
+            IntConstantValue::I8(val) => Scalar::from_int(val, size),
+            IntConstantValue::I16(val) => Scalar::from_int(val, size),
+            IntConstantValue::I32(val) => Scalar::from_int(val, size),
+            IntConstantValue::I64(val) => Scalar::from_int(val, size),
+            IntConstantValue::I128(val) => Scalar::from_int(val, size),
+            IntConstantValue::U8(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U16(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U32(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U64(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U128(val) => Scalar::from_uint(val, size),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Constructor)]
+pub struct ScalarInt {
+    /// The underlying scalar value.
+    data: Scalar,
+
+    /// Whether the integer is signed or not.
+    ty: IntTy,
+}
+
+impl fmt::Display for ScalarInt {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let size = self.data.size();
+        let bits = self.data.to_bits(size).unwrap();
+
+        if self.ty.is_signed() {
+            let value = size.sign_extend(bits) as i128;
+            write!(f, "{}_{}", value, self.ty.normalise(size))
+        } else {
+            write!(f, "{}_{}", bits, self.ty.normalise(size))
+        }
+    }
+}
+
+impl From<f32> for Scalar {
+    fn from(value: f32) -> Self {
+        Self {
+            value: value.to_bits() as u128,
+            size: NonZeroU8::new(std::mem::size_of::<f32>() as u8).unwrap(),
+        }
+    }
+}
+
+impl TryFrom<Scalar> for f32 {
+    type Error = Size;
+
+    fn try_from(value: Scalar) -> Result<Self, Self::Error> {
+        value
+            .to_bits(Size::from_bytes(std::mem::size_of::<f32>()))
+            .map(|u| f32::from_bits(u as u32))
+    }
+}
+
+impl From<f64> for Scalar {
+    fn from(value: f64) -> Self {
+        Self {
+            value: value.to_bits() as u128,
+            size: NonZeroU8::new(std::mem::size_of::<f64>() as u8).unwrap(),
+        }
+    }
+}
+
+impl TryFrom<Scalar> for f64 {
+    type Error = Size;
+
+    fn try_from(value: Scalar) -> Result<Self, Self::Error> {
+        value
+            .to_bits(Size::from_bytes(std::mem::size_of::<f64>()))
+            .map(|u| f64::from_bits(u as u64))
+    }
+}
+
+/// An [Alloc] represents a single allocation slot which contains the actual
+/// data, alignment, and mutability of the allocation.
+#[derive(Clone, Debug)]
+pub struct Alloc<Buf = Box<[u8]>> {
```

Comment:
One potential thing to do is to use this for some TIR arrays. E.g. maybe 
```
enum ArrayTerm {
   Concrete(Alloc),
   Symbolic(TermListId)
}
```
Some more thought is needed for this.

Reply:
Yeah perhaps it could use the `Alloc`. However, this would mean that the Alloc/Scalar representations (which sometimes use type info) would have to be type agnostic. 

Since it would need to understand both `IrTy`s and `Ty`s. The simple way would be to convert the TIR type into an IR type? If we do this, we'll need to do some minor restructuring of the crates to avoid cyclic deps.

Reply:
Ideally, `hash-ir` shouldn't depend on `hash-tir`, and the alloc/scalar stuff is moved into something like `hash-source`. Then, `hash-ir-utils` could provide the tools to convert between type representations, and interface with the alloc API (as it does now)?